### PR TITLE
Added support for reading ELF files.

### PR DESCRIPTION
**Pull request by simonjwinwood**

### CODE REVIEW

File: lean/elf.lean [hunk 1]
```diff
@@ -322,29 +325,85 @@ end word
...
-(val : uint32)
+inductive phdr_type : Type
+  | PT_NULL    : phdr_type
+  | PT_LOAD    : phdr_type
+  | PT_DYNAMIC : phdr_type
+  | PT_INTERP  : phdr_type
+  | PT_NOTE    : phdr_type 
+  | PT_SHLIB   : phdr_type
+  | PT_PHDR    : phdr_type
+  | PT_PROC    : uint32 -> phdr_type
```
Comment (joehendrix) [inline lean/elf.lean]: I'm inclined to keep the original definition for this, and flags, but add definitions for constants e.g.,

  `def PT_NULL : phdr_type := \{0\}`
 
We might then use an rb_map to match phdr types to the string.

File: lean/elf.lean [hunk 2]
```diff
@@ -531,6 +590,27 @@ def pp_phdrs {c:elf_class} (phdrs:list (phdr c)) : io unit := do
...
+def read_one_elfmem {c : elf_class} (m : elfmem) (ph : phdr c) : file_input elfmem :=
+  match ph.phdr_type with
+  | phdr_type.PT_LOAD := do
+    file_input.seek ph.offset.val,
+    fbs <- file_input.read (min ph.filesz.val ph.memsz.val),
+    monad_lift (io.put_str_ln ("read_one_elfmem: read " ++ fbs.size.repr ++ " bytes")),
+    let val := buffer.append_list fbs (list.repeat (char.of_nat 0) (ph.memsz.val - ph.filesz.val)) 
+    in pure $ data.imap.insert ph.vaddr.val ph.memsz.val val m
+  | _       := pure m
+  end
```
Comment (joehendrix) [inline lean/elf.lean]: Could you add a comment on what this and region/elfmem definitions are intended to do?

File: lean/file_input.lean [hunk 3]
```diff
@@ -0,0 +1,36 @@
+
+import system.io -- init.category.lift
+
+/- A monad for dealing with the fact that lean doesn't support seeking backwards -/
```
Comment (joehendrix) [inline lean/file_input.lean]: Should comment be moved from `file_input_state` to `file_input`?

File: lean/imap.lean [hunk 4]
```diff
@@ -0,0 +1,33 @@
+/- A simple interval map -/
+
+structure {u v} data.imap.imap_entry (k : Type u) (val : Type v) : Type (max u v) := 
+  (start : k) 
+  (extent : k)
+  (value : val)
+
+def {u v} data.imap (k : Type u) (val : Type v) := list (data.imap.imap_entry k val)
```
Comment (joehendrix) [inline lean/imap.lean]: Shouldn't ordering be part of type for this and `imap_entry` as in `rb_map`?

Comment (joehendrix) [reply lean/imap.lean]: Test comment

File: lean/elf.lean (hunk 1 above)
Comment (simonjwinwood) [reply lean/elf.lean]: I can revert the change, but am interested to know why you prefer a uint32 to the sum type?

File: lean/imap.lean (hunk 4 above)
Comment (simonjwinwood) [reply lean/imap.lean]: I added the lt param to the type of imap, but we might want to figure out whether we want some more complicated datastructure before following rbtree/rbmap

File: lean/elf.lean (hunk 1 above)
Comment (joehendrix) [reply lean/elf.lean]: The `uint32` is compact and isomorphic to the value stored on disk.  It also treats the OS/processor specific values uniformly as predefined ones.

**Comment by simonjwinwood** [comment]
Superseded by #9

**Pull request closed without merging**